webhook/resourcesemantics/conversion/controller.go

```diff
@@ -0,0 +1,130 @@
+// are required to implement in order to be convertible from
+// one version to another
+//
+// Optionally if the object implements apis.Defaultable the
+// ConversionController will apply defaults before returning
+// the response
+type ConvertibleObject interface {
+	// ConvertUp(ctx)
+	// ConvertDown(ctx)
+	apis.Convertible
+
+	// DeepCopyObject()
+	// GetObjectKind() => SetGroupVersionKind(gvk)
+	runtime.Object
+}
+
+// GroupKindConversion specifies how a specific Kind for a given
+// group should be converted
+type GroupKindConversion struct {
+	// DefinitionName specifies the CustomResourceDefinition that should
+	// be reconciled with by the controller.
+	//
+	// The conversion webhook configuration will be updated
+	// when the CA bundle changes
+	DefinitionName string
+
+	// HubVersion specifies which version of the CustomResource supports
+	// convertions to and from all types
+	//
+	// It is expected that the Zygotes map contains an entry for the
+	// specified HubVersion
+	HubVersion string
+
+	// Zygotes contains a map of version strings (ie. v1, v2) to empty
+	// ConvertibleObject objects
+	//
+	// During a conversion request these zygotes will be deep copied
+	// and manipulated using the apis.Convertible interface
+	Zygotes map[string]ConvertibleObject
+}
+
+// NewConversionController returns a K8s controller that will
+// will reconcile CustomResourceDefinitions and update their
+// conversion webhook attributes such as path & CA bundle.
+//
+// Additionally the controller's Reconciler implements
+// webhook.ConversionController for the purposes of converting
+// resources between different versions
+func NewConversionController(
+	ctx context.Context,
+	path string,
+	kinds map[schema.GroupKind]GroupKindConversion,
+	withContext func(context.Context) context.Context,
+) *controller.Impl {
+
+	logger := logging.FromContext(ctx)
+	secretInformer := secretinformer.Get(ctx)
+	crdInformer := crdinformer.Get(ctx)
+	client := apixclient.Get(ctx)
+	options := webhook.GetOptions(ctx)
+
+	r := &reconciler{
+		kinds:       kinds,
+		path:        path,
+		secretName:  options.SecretName,
+		withContext: withContext,
+
+		client:       client,
+		secretLister: secretInformer.Lister(),
+		crdLister:    crdInformer.Lister(),
+	}
+
+	c := controller.NewImpl(r, logger, "ConversionWebhook")
+
+	// Reconciler when the named CRDs change.
+	for _, gkc := range kinds {
+		name := gkc.DefinitionName
+
+		crdInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+			FilterFunc: controller.FilterWithName(name),
+			Handler:    controller.HandleAll(c.Enqueue),
+		})
+
+		sentinel := c.EnqueueSentinel(types.NamespacedName{Name: name})
+
+		// Reconcile when the cert bundle changes.
+		secretInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+			FilterFunc: controller.FilterWithNameAndNamespace(system.Namespace(), options.SecretName),
+			Handler:    controller.HandleAll(sentinel),
+		})
```

Comment:
This doesn't need to be in the loop

Reply:
It does.

When the secret changes we need to reconcile all the CRDs we're tracking in order to update their webhook config. So the sentinel value actually changes on line 128

Reply:
I missed that the Sentinel is changing, which is a bit weird.  Generally I'd expect `EnqueueSentinel` to be used with singleton controllers that aren't paying attention to the key at all (we probably shouldn't have even exposed a parameter, which in hindsight was probably a mistake).

Should this be using EnqueueKey instead of EnqueueSentinel?

Reply:
> Should this be using EnqueueKey instead of EnqueueSentinel?

I think `EnqueueSentinel` still fits since we're reconciling a fixed set of objects (CRDs).

Reply:
The spirit of `EnqueueSentinel` was to trigger a reconciliation where the key that's enqueued doesn't matter.  `EnqueueKey` should be used when what's enqueued matters, whether it's a discrete or unbounded set.

Reply:
I'm confused - in the usage actioning on secret updates and enqueuing the fixed list of CRD keys to be reconciled. Passing `EnqueueKey` would enqueue the secret's key.

Reply:
Unless you're suggesting write a custom informer handler where I invoke `EnqueueKey` N-times for the CRD resources we're watching.

Reply:
I see now.  I guess the closest analog would be a global resync today, but because of the fixed keyspace you are able to do it with a fixed N informer watches.

I think I still have a mild preference for `EnqueueKey` with a fixed key over `EnqueueSentinel`, but this does end up looking a bit grosser due to the need for a thunk (if I'm not mistaken, I certainly need more ☕️ )